src/test/storage.zig

```diff
@@ -456,33 +544,280 @@ pub const Storage = struct {
+
+            const target_area = SuperBlockArea.from_offset(offset_in_zone);
+            // This is the maximum number of faults per-area that can be safely injected on a read
+            // or write to the superblock zone.
+            //
+            // For SuperBlockSector, checkpoint() and view_change() require 3/4 valid sectors (1
+            // fault). Trailers are likewise 3/4 + 1 fault — consider if two faults were injected:
+            // 1. `SuperBlock.checkpoint()` for sequence=6.
+            //   - write copy 0, corrupt manifest (fault_count=1)
+            //   - write copy 1, corrupt manifest (fault_count=2) !
+            // 2. Crash. Recover.
+            // 3. `SuperBlock.open()`. The highest valid quorum is sequence=6, but there is no
+            //    valid manifest.
+            const fault_count_max = @divExact(config.superblock_copies, 2) - 1;
+            assert(fault_count_max >= 1);
+
+            const fault_count = blk: {
+                var fault_count: usize = 0;
+                var copy_: u8 = 0;
+                while (copy_ < config.superblock_copies) : (copy_ += 1) {
+                    const copy_area = SuperBlockArea{ .group = target_area.group, .copy = copy_ };
+                    const copy_area_offset_zone = copy_area.to_offset();
+                    const copy_area_offset_storage = zone.offset(copy_area_offset_zone);
+                    const copy_area_sector = @divExact(copy_area_offset_storage, config.sector_size);
+                    fault_count += @boolToInt(storage.faults.isSet(copy_area_sector));
+                }
+                break :blk fault_count;
+            };
+
+            // fault_count may be slightly greater than fault_count_max due to faults added by
+            // `Storage.reset()` (a simulated crash).
+            assert(fault_count <= fault_count_max + 1);
+            if (fault_count >= fault_count_max) return null;
+
+            // Always fault the first sector of the read/write so that we can easily test
+            // `storage.faults` to probe the current `fault_count`.
+            const sector = @divExact(offset_in_storage, config.sector_size);
+            return SectorRange{
+                .min = sector,
+                .max = sector + 1,
+            };
+        }
 
-        const start = std.math.max(offset, faulty_offset);
-        const end = std.math.min(offset + size, faulty_offset + message_size_max);
+        if (zone == .wal_headers or zone == .wal_prepares) {
+            const faulty_wal_areas = storage.options.faulty_wal_areas orelse return null;
+            const message_size_max = config.message_size_max;
+            const period = faulty_wal_areas.period;
 
-        // The read/write does not touch any faulty sectors.
-        if (start >= end) return null;
+            const offset_faulty =
+                faulty_wal_areas.first_offset + (offset_in_storage / period) * period;
```

Comment:
```suggestion
                faulty_wal_areas.first_offset + @divTrunc(offset_in_storage, period) * period;
```

A little more obvious.

Reply:
I kind of hope that zig removes integer division via `/` entirely in favor of explicit use of `@divTrunc` etc.

Reply:
I try to avoid `/` when I remember, but it isn't automatic for me yet.